### PR TITLE
feat(metrics): Tensorboard support.

### PR DESCRIPTION
**Pull request by weikaolun**
This PR adds a functional `sync_tensorboard` option to the `metrics.init()` call, and changes the semantics of that call to leverage `with`. This implementation parses raw numeric values from Tensorboard events and logs them as metrics in Gradient. Currently, tensor-encoded numeric data is not supported.

### CODE REVIEW

File: gradient_utils/metrics/metrics.py [hunk 1]
```diff
@@ -6,19 +6,43 @@
...
+            labelnames=[
+                get_workload_label(),
+                "pod",
+                "step"])
+        new_metric = new_metric_raw.labels(get_workload_id(), HOSTNAME, step)
+        new_metric.set(metric.value)
+        new_grouping_key = {
+            "name": metric.key,
+        }
+        new_grouping_key.update(default_grouping_key)
```
Comment (lamroger) [inline gradient_utils/metrics/metrics.py:37]: i think you need to pass step into default_grouping_key

Maybe default_grouping_key should be first and then name and step should be updated. this way feels more natural

File: gradient_utils/metrics/tensorboard/patch.py [hunk 2]
```diff
@@ -93,5 +100,116 @@ def __init__(self, *args, **kwargs):
...
-def on_new_logdir(logdir):
-    logger.debug(logdir)
+def on_new_logdir(logdir, settings):
+    logger.debug("watching %s", logdir)
+    watcher = LogdirWatcher(logdir, settings)
+    settings.tensorboard_watchers.append(watcher)
+    watcher.start()
+
+
+class LogdirWatcher(object):
```
Comment (lamroger) [inline gradient_utils/metrics/tensorboard/patch.py:110]: `object` is redundant in Python3 (apparently)

File: gradient_utils/metrics/tensorboard/patch.py [hunk 3]
```diff
@@ -93,5 +100,116 @@ def __init__(self, *args, **kwargs):
...
+                logger.debug("watch stopped")
+                break
+
+    def _process_event(self, event):
+        if event.HasField("summary"):
+            try:
+                step = event.step
+                summary = event.summary
+                for value in summary.value:
+                    name = value.tag.replace(".", "_").replace("/", "_").replace(" ", "_")
```
Comment (lamroger) [inline gradient_utils/metrics/tensorboard/patch.py:145]: Could you add a small comment on why

**Review by lamroger — COMMENTED** [review]
small comments - looks g

File: gradient_utils/metrics/tensorboard/patch.py (hunk 2 above)
Comment (weikaolun) [reply gradient_utils/metrics/tensorboard/patch.py:110]: have we officially dropped support for python2 in this repo?

File: gradient_utils/metrics/metrics.py (hunk 1 above)
Comment (weikaolun) [reply gradient_utils/metrics/metrics.py:37]: it's passed into the call to `_get_default_grouping_key` on line 21.

File: gradient_utils/metrics/tensorboard/patch.py (hunk 2 above)
Comment (lamroger) [reply gradient_utils/metrics/tensorboard/patch.py:110]: Yea, I did when I revamped the build system bc Python2 is EOL

**Review by lamroger — DISMISSED** [review]

**Review by lamroger — APPROVED** [review]

**Pull request merged**